java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -191,6 +355,74 @@ public void pruneRow(Row key, long lowerBound) throws Exception {
                 timestampToRow.remove(cur);
                 iterator.remove();
             }
+
+            Row accumulators = null;
+            final Map<String, Integer> leftIdxMap = getLeftTimestampIdx();
+            for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
+                    aggFieldsDescriptor.getAggFieldDescriptors()) {
+                final int removedRowCnt = originalSize - timestamps.size();
+                final int leftIdx =
+                        leftIdxMap.getOrDefault(descriptor.outFieldName, 0) - removedRowCnt;
+                if (leftIdx <= 0) {
+                    leftIdxMap.remove(descriptor.outFieldName);
+                } else {
+                    leftIdxMap.put(descriptor.outFieldName, leftIdx);
+                }
+                if (leftIdx >= timestamps.size() || timestamps.get(leftIdx) > currentTimestamp) {
+                    if (accumulators == null) {
+                        accumulators = getAccumulators();
+                    }
+                    accumulators.setField(descriptor.outFieldName, null);
+                }
+            }
+            updateLeftTimestampIdx(leftIdxMap);
+            if (accumulators != null) {
+                updateAccumulators(accumulators);
+            }
+
+            if (timestamps.isEmpty()) {
+                timestampList.clear();
+                accumulatorsRow.clear();
+            } else {
+                timestampList.update(timestamps);
```

Comment:
Suppose window_size=24 hours and step_size=1 sec, then we need to overwrite a list of size 24 * 60 * 60 per second per key.

And alternative solution is to reduce the frequency of removing entries from the list in this timestampList. For example, we can update it only once every window_size/step_size/10 steps. In the case described above, it means we will only overwrite the entire list every 2.4 hours. The cost this approach is that the list size will be 10% more than the existing approach. But it seems like a good tradeoff to make. What do you think?

Reply:
You are right that we are updating the timestamp list too frequently. Take it one step further, I think we don't need to prune rows at every call to the `onTimer`. 

The approach to only pruneRow every x step may not work because it is not guaranteed that an input row for each key will come at every step size. This can lead to some very old rows not getting pruned if the input rows of a key are coming infrequently. 

So I'd like to introduce a prune row grace period that specifies the maximum duration a row can stay in the state after it is out of all windows. And we set the grace period to be 1/10 of the max window size so that the state size is
at most 1/10 larger and we only need to prune rows once every 1/10 of the maximum window time.

Reply:
Sounds good. Actually this is the same approach suggested in the first comment.